aws-java-sdk-core/src/main/java/com/amazonaws/internal/SdkThreadLocalsRegistry.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class SdkThreadLocalsRegistry {
+    
+    private static final List<ThreadLocal<?>> threadLocals = new ArrayList<ThreadLocal<?>>();
+
+    private SdkThreadLocalsRegistry() {
+        // prevent instantiation
+    }
+    
+    public static synchronized <T> ThreadLocal<T> register(ThreadLocal<T> threadLocal) {
+        threadLocals.add(threadLocal);
+        return threadLocal;
+    }
```

Comment:
Right now, if `register` is used from a non-static final context, we could leak memory. Or if it's used on a critical path, we could be creating a performance bottleneck with the `synchronized` methods.

An alternative (more complex) implementation could fix those issues, if we see them as a problem:

```Java
class SdkThreadLocalsRegistry {
    private static final ThreadLocal<Set<WeakReference<ThreadLocal<?>>>> REGISTRY =
            ThreadLocal.withInitial(HashSet::new);
    private static final ThreadLocal<ReferenceQueue<Object>> RELEASED_REFERENCE_QUEUE =
            ThreadLocal.withInitial(ReferenceQueue::new);

    public static void register(ThreadLocal<?> threadLocal) {
        Set<WeakReference<ThreadLocal<?>>> threadLocalRegistry = REGISTRY.get();
        clearExpiredEntries(threadLocalRegistry);
        threadLocalRegistry.add(new WeakReference<>(threadLocal, RELEASED_REFERENCE_QUEUE.get()));
    }

    public static void remove() {
        REGISTRY.get().stream()
                .map(WeakReference::get)
                .filter(l -> l != null)
                .forEach(ThreadLocal::remove);
        REGISTRY.remove();
        RELEASED_REFERENCE_QUEUE.remove();
    }

    private static void clearExpiredEntries(Set<WeakReference<ThreadLocal<?>>> threadLocalRegistry) {
        Reference<?> releasedReference;
        while ((releasedReference = RELEASED_REFERENCE_QUEUE.get().poll()) != null) {
            threadLocalRegistry.remove(releasedReference);
        }
    }
}
```

This might introduce new issues I'm not aware of...

Reply:
At the moment `register` will be called a maximum of 5 times because the ThreadLocal instances are class level singletons and only on creation of those singletons is `register` called. 

I think contention on the `SdkThreadLocals.remove()` method from multiple threads is unlikely because it would normally be called after a batch of single threaded IO interaction with AWS which will naturally limit the maximum call rate from a single thread. Perhaps I'm not being imaginative enough with expected usage?

Bear in mind that any addition of a ThreadLocal to the code base will have to trigger reviewers to suggest it be registered and at that point contention issues can be revisited.  I think the suggested replacement code adds unnecessary complexity for the moment given that a concrete use case (ThreadLocal instantiation that is not a class level singleton) has not been proposed for the sdk. The code you are suggesting may never be needed.

Having said all that, that's a nifty pattern you've suggested and I'm happy to review and include if desired.

Reply:
> It's probably good to add a comment that register should only be used from static final locations (eg. finals in enums and static final variables).

Will do.

> Is this method used cross-module? If so, it would be good to mark it as an @SdkProtectedApi .

It's used in *aws-java-sdk-core* and *aws-java-sdk-sns* which I assume means it's cross module.  I'll add that annotation.